File: codesite/python_problems/utils/linked_list_utils.py
from typing import Optional, List


class ListNode:
    """
    Definition for singly-linked list.
    """

    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


def build_linked_list(nums: List[int], cycle_position: int = -1) -> ListNode:
    """
    Build linked list with cycle from list.
    """
    anchor = node = ListNode()
    cycle_node = None

    for position, num in enumerate(nums):
        node.next = ListNode(num)
        node = node.next

        if position == cycle_position:
            cycle_node = node

    if cycle_node is not None:
        node.next = cycle_node

    return anchor.next


def serialize_linked_list(root: ListNode) -> List[int]:
    """
    Return linked list values in list.
    """
    values = []
    while root:
        values.append(root.val)
        root = root.next
    return values


def are_lists_equeal(root1: ListNode, root2: ListNode) -> bool:
    """
    Compare two linked lists value by value.
    """
    node1 = root1
    node2 = root2

    while node1 or node2:
        if node1 is None and node2 is None:
            return True
        elif (
            node1 is None or node2 is None or
            node1.val != node2.val
        ):
            return False

        node1 = node1.next
        node2 = node2.next
    return True

File: codesite/python_problems/utils/linked_list_utils.cpp
#include <vector>
using namespace std;


struct ListNode {
   int val;
   ListNode* next;
   ListNode() : val(-1), next(nullptr) {}
   ListNode(int val) : val(val), next(nullptr) {}
   ListNode(int val, ListNode* next) : val(val), next(next) {}
};


struct LinkedList {
   ListNode* buildLinkedList(const vector<int> nums, int cyclePosition = -1) {
      ListNode* node = new ListNode();
      ListNode* anchor = node;
      ListNode* cycleNode = nullptr;

      for (int position = 0; position < nums.size(); ++position) {
         int num = nums[position];
         node->next = new ListNode(num);
         node = node->next;

         if (position == cyclePosition) {
            ListNode* cycleNode = node;
         }
      }

      if (cycleNode) {
         node->next = cycleNode;
      }

      return anchor->next;
   }

   vector<int> serializeLinkedList(ListNode* node) {
      vector<int> values;

      while (node) {
         values.push_back(node->val);
         node = node->next;
      }

      return values;
   }

   bool areLinkedListsEqueal(ListNode* root1, ListNode* root2) {
      ListNode* node1 = root1;
      ListNode* node2 = root2;

      while (node1 || node2) {
         if (node1 == nullptr && node2 == nullptr) {
            return true;
         }
         else if (
            (node1 == nullptr || node2 == nullptr)
            || node1->val != node2->val
            ) {
            return false;
         }

         node1 = node1->next;
         node2 = node2->next;
      }

      return true;
   }
};

